src/utils/calendar.ts

```diff
@@ -0,0 +1,34 @@
+import { PersonalizedVenueEvent } from "types/venues";
+import { VenueEvent } from "types/venues";
+import { eventEndTime, eventStartTime } from "utils/event";
+import { WithVenueId } from "utils/id";
+import { getFullVenueInsideUrl } from "utils/url";
+
+import ical from "ical-generator";
+
+export const createCalendar = (
+  allEvents: PersonalizedVenueEvent[] | WithVenueId<VenueEvent>[],
+  calendarname: string
+) => {
+  const cal = ical({ name: calendarname });
+
+  if (!cal) return;
+
+  allEvents.map((event: PersonalizedVenueEvent | WithVenueId<VenueEvent>) =>
+    cal.createEvent({
+      id: event.venueId,
+      start: eventStartTime(event),
+      end: eventEndTime(event),
+      description: event.host,
+      summary: event.name ? event.host : "",
+      url: getFullVenueInsideUrl(event.venueId),
+    })
+  );
+
+  const outputCal = cal.toURL();
+  const outputFile = `${calendarname}.ics`;
+  const link = document.createElement("a");
+  link.download = outputFile;
+  link.href = outputCal;
+  link.click();
```

Comment:
This part probably belongs in a different function (maybe called `downloadCalendar`). `createCalendar` should just create and `return` the calendar object (with the added events).

We could have a new function that accepts a calendar object, and then triggers this download (`downloadCalendar`)

Reply:
See also the discussion in https://github.com/sparkletown/sparkle/pull/1423#issuecomment-848630538 for what the `cal.toURL()` produces here, and whether we can make that work with a `webcal://` prefix or similar.

Further Reading:

- https://sebbo2002.github.io/ical-generator/develop/reference/classes/icalcalendar.html#tourl
- https://sebbo2002.github.io/ical-generator/develop/reference/classes/icalcalendar.html#toblob
- https://developer.mozilla.org/en-US/docs/Web/API/Blob
- https://developer.mozilla.org/en-US/docs/Web/API/Blob#creating_a_url_representing_the_contents_of_a_typed_array
- https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL
- https://developer.mozilla.org/en-US/docs/Web/API/File/Using_files_from_web_applications#using_object_urls

Reply:
Not adapting this solution see ec68e83